File: go/test/inbox_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/tempmail-sdk/go"
	"github.com/voxgig-sdk/tempmail-sdk/go/core"

	vs "github.com/voxgig-sdk/tempmail-sdk/go/utility/struct"
)

func TestInboxEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Inbox(nil)
		if ent == nil {
			t.Fatal("expected non-nil InboxEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := inboxBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "inbox." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set TEMPMAIL_TEST_INBOX_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		inboxRef01Ent := client.Inbox(nil)
		inboxRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "inbox"}), "inbox_ref01"))
		inboxRef01Data["domain"] = setup.idmap["domain01"]
		inboxRef01Data["username"] = setup.idmap["username01"]

		inboxRef01DataResult, err := inboxRef01Ent.Create(inboxRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		inboxRef01Data = core.ToMapAny(entityData(inboxRef01DataResult))
		if inboxRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}

		// LOAD
		inboxRef01MatchDt0 := map[string]any{}
		inboxRef01DataDt0Loaded, err := inboxRef01Ent.Load(inboxRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if inboxRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func inboxBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "inbox", "InboxTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read inbox test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse inbox test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"inbox01", "inbox02", "inbox03", "domain01", "username01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("TEMPMAIL_TEST_INBOX_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"TEMPMAIL_TEST_INBOX_ENTID": idmap,
		"TEMPMAIL_TEST_LIVE":      "FALSE",
		"TEMPMAIL_TEST_EXPLAIN":   "FALSE",
		"TEMPMAIL_APIKEY":         "",
	})

	idmapResolved := core.ToMapAny(env["TEMPMAIL_TEST_INBOX_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["TEMPMAIL_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["TEMPMAIL_APIKEY"],
			},
			extraOpts,
		})
		client = sdk.NewTempmailSDK(core.ToMapAny(mergedOpts))
	}

	live := env["TEMPMAIL_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["TEMPMAIL_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/InboxEntityTest.php
<?php
declare(strict_types=1);

// Inbox entity test

require_once __DIR__ . '/../tempmail_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class InboxEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = TempmailSDK::test(null, null);
        $ent = $testsdk->Inbox(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = inbox_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create", "load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "inbox." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set TEMPMAIL_TEST_INBOX_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $inbox_ref01_ent = $client->Inbox(null);
        $inbox_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.inbox"), "inbox_ref01"));
        $inbox_ref01_data["domain"] = $setup["idmap"]["domain01"];
        $inbox_ref01_data["username"] = $setup["idmap"]["username01"];

        [$inbox_ref01_data_result, $err] = $inbox_ref01_ent->create($inbox_ref01_data, null);
        $this->assertNull($err);
        $inbox_ref01_data = Helpers::to_map($inbox_ref01_data_result);
        $this->assertNotNull($inbox_ref01_data);

        // LOAD
        $inbox_ref01_match_dt0 = [];
        [$inbox_ref01_data_dt0_loaded, $err] = $inbox_ref01_ent->load($inbox_ref01_match_dt0, null);
        $this->assertNull($err);
        $this->assertNotNull($inbox_ref01_data_dt0_loaded);

    }
}

function inbox_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/inbox/InboxTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = TempmailSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["inbox01", "inbox02", "inbox03", "custom01", "custom02", "custom03", "domain01", "username01"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("TEMPMAIL_TEST_INBOX_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "TEMPMAIL_TEST_INBOX_ENTID" => $idmap,
        "TEMPMAIL_TEST_LIVE" => "FALSE",
        "TEMPMAIL_TEST_EXPLAIN" => "FALSE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["TEMPMAIL_TEST_INBOX_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["TEMPMAIL_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
            ],
            $extra ?? [],
        ]);
        $client = new TempmailSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["TEMPMAIL_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["TEMPMAIL_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
